Add more STMOD+ PINs description and review power-up sequence

LMESTM opened the pull request:
Also add STMOD+ cellular type to select either BG96 or UG96.
BG96 supported by default.

Tested OK with mbed-os-example-cellular with both PPP mode true and false.

LMESTM committed:
Add more STMOD+ PINs description and review power-up sequence

Also add STMOD+ cellular type to select either BG96 or UG96.
BG96 supported by default.

LMESTM commented:
@ithinuel  Here we are with a new version - tested ok for now on BG96 + PPP disabled

ithinuel commented on STModCellular.cpp:
```diff
@@ -18,53 +18,104 @@
+    tr_debug("STModCellular creation\r\n");
+
+    /* Ensure PIN SIMs are set as input */
+    DigitalIn sim_reset(MBED_CONF_STMOD_CELLULAR_SIM_RESET);
```
I don't think the state is guaranteed once the object gets destroyed.
These shall probably be members.

LMESTM replied:
good point ! will fix

ithinuel commented on STModCellular.cpp:
```diff
@@ -18,53 +18,104 @@
+    m_powerkey.write(0);
+    /* Because modem status is not available on STMOD+ connector,
+     * let's wait for Modem complete boot */
+    wait_ms(2300);
```
Is this `wait_ms` required since we are waiting for the RDY signal ?

LMESTM replied:
I think it is required - I understand that reading UART before this timing may be unpredictable.  This is the reliable sequence that I was given to and should be followed (As stated in the comment, if modem_status pin would be available, we would use this signal before reading uart, but it isn't)

ithinuel commented on STModCellular.cpp:
```diff
@@ -18,53 +18,104 @@
+#endif
+#if (MBED_CONF_STMOD_CELLULAR_TYPE == STMOD_BG96)
+    tr_debug("Booting BG96\r\n");
+    m_powerkey.write(1);
```
Why is the powerkey toggled while the device is under reset ? Is that a required for this module ?

LMESTM replied:
That is the recommended "Turn on Module Using the PWRKEY Pin" from the BG96 HW design doc. 
We ensure both RESET and "No Power", then we release Power and reset at the same time, then toggle the PWRKEY pin and wait for 5sec (actually spec says 4,9sec but well  ... )

ithinuel commented on STModCellular.cpp:
```diff
@@ -18,53 +18,104 @@
+    /*  Modem may send more bytes are RDY flag */
+    _at->flush();
+
+    /* Turn OFF ECHO before anything else */
```
Is this required here since it is also done by the `CellularStateMachine` ?

LMESTM replied:
We want to turn it off as soon as possible, before Flow control commands, otherwise we're getting RX characters in AT fifo that seem to be not well parsed and may lead to error later on ... I've seen it more stable when doing so.

ithinuel commented on STModCellular.cpp:
```diff
@@ -73,37 +124,50 @@ nsapi_error_t STModCellular::soft_power_on() {
             tr_error("Failed to enable hw flow control");
         }
-
-        _at->restore_at_timeout();
```
Moving the `unlock()` about the `if` statement seems fine for me but I think we still need to `restore_at_timeout()` before releasing the lock.

LMESTM replied:
Right - Done

ithinuel commented on STModCellular.cpp:
```diff
@@ -73,37 +124,50 @@ nsapi_error_t STModCellular::soft_power_on() {
+    wait_ms(500);
+
+    _at->lock();
+    /*  Verify Flow Control settings */
```
Are we actually verifying anything here ? Maybe shall this be done only if `AT_DEBUG` is set ?

LMESTM replied:
Makes sense, will do !

ithinuel commented on STModCellular.cpp:
```diff
@@ -73,37 +124,50 @@ nsapi_error_t STModCellular::soft_power_on() {
+    tr_debug("MODEM default instance\r\n");
+
     static UARTSerial serial(MBED_CONF_STMOD_CELLULAR_TX, MBED_CONF_STMOD_CELLULAR_RX, MBED_CONF_STMOD_CELLULAR_BAUDRATE);
+#if defined (MBED_CONF_STMOD_CELLULAR_RTS) && defined(MBED_CONF_STMOD_CELLULAR_CTS)
```
This shall actually be done after the `AT+IFC` command is issued. Otherwise we may endup in a locked state where the mcu waits for the modem to be ready to receive while it is actually a floating `CTS` pin.

LMESTM replied:
Hmm actually we need to initialize the UART before we start the UART transmissions ... 
this seems also to be done this way for all other modems in mbed-os tree:
https://github.com/ARMmbed/mbed-os/blob/master/features/cellular/framework/targets/QUECTEL/BG96/QUECTEL_BG96.cpp#L82

ithinuel replied:
If `AT+IFC` follows the same logic as `AT+IPR`, according to [Quectel's](https://www.quectel.com/UploadImage/Downlad/Quectel_BG96_AT_Commands_Manual_V2.1.pdf#%5B%7B%22num%22%3A107%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C51%2C621%2C0%5D) documentation, the configuration takes place after any result code is delivered.

All other modem's driver may suffer from the same issue ;-)

LMESTM replied:
> All other modem's driver may suffer from the same issue ;-)
I understand the point.  Still I haven't seen any issue with this sequence, and I guess developers of other modem drivers have checked the same. 

Anyway, I will git it a try and see if we can activate flow control after OK response.

ithinuel reviewed (changes requested):
Hi @LMESTM thanks for your update.
I left few reviews, let me know what are your thoughts about them. :)

LMESTM committed:
Move sim pins to object members

LMESTM committed:
Only check Flow Control setting if DEBUG AT is on

This is just a debug check, we're not parsing the AT answer.
So the check would not be visible if no AT debug.

LMESTM committed:
Restore at timeout after Flow Control setting

LMESTM commented:
@ithinuel  I have pushed a branch with several fixes and I have posted an answer to other comments.

jeromecoutant commented on mbed_lib.json:
```diff
@@ -54,10 +70,13 @@
         "DISCO_L496AG": {
             "tx": "PB_6",
```
Could we use STMOD_2 name ?
https://github.com/ARMmbed/mbed-os/blob/46603f831e13705d5aab4f73d52f00611e1f4d7d/targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L496xG/TARGET_DISCO_L496AG/PinNames.h#L354

ithinuel replied:
That sounds sensible to me.

LMESTM replied:
excellent idea - will make the driver configuration valid for any target with STMOD+ connector

LMESTM committed:
Define STMOD default pins with STMOD definition

This makes default pin map defined for any STM target that supports
the STMOD+ connector and have the correct STMOD_XX defines.
